src/platform/packages/shared/kbn-esql-ast/src/definitions/utils/validation/function.ts

```diff
@@ -6,930 +6,317 @@
+    for (const arg of this.fn.args) {
+      this.argTypes.push(
+        getExpressionType(arg, this.context.fields, this.context.userDefinedColumns)
       );
-    }
-    if (messages.length) {
-      return messages;
+      this.argLiteralsMask.push(isLiteral(arg));
     }
   }
 
-  const matchingSignatures = getSignaturesWithMatchingArity(fnDefinition, fn);
+  validate(): ESQLMessage[] {
+    const nestedErrors = this.validateNestedFunctions();
 
-  // Check license requirements for the function
-  const hasMinimumLicenseRequired = callbacks?.hasMinimumLicenseRequired;
-
-  if (isFnSupported.supported && hasMinimumLicenseRequired) {
-    const licenseMessages = validateFunctionLicense(fn, hasMinimumLicenseRequired);
-    if (licenseMessages.length) {
-      messages.push(...licenseMessages);
+    if (nestedErrors.length) {
+      // if one or more child functions produced errors, stop validation and report
+      return nestedErrors;
     }
 
-    // Check signature-specific license requirements
-    if (licenseMessages.length === 0 && matchingSignatures.length > 0) {
-      const signatureLicenseMessages = validateSignatureLicense(
-        fn,
-        matchingSignatures,
-        hasMinimumLicenseRequired,
-        context,
-        parentCommand
-      );
+    if (isParamLiteral(this.fn.operator)) {
+      // skip validation for functions with names defined by a parameter
+      // e.g. "... | EVAL ??param(..args)"
+      return [];
+    }
 
-      if (signatureLicenseMessages.length) {
-        messages.push(...signatureLicenseMessages);
-      }
+    if (!this.definition) {
+      return [errors.unknownFunction(this.fn)];
     }
-  }
 
-  if (!matchingSignatures.length) {
-    const { max, min } = getMaxMinNumberOfParams(fnDefinition);
-    if (max === min) {
-      messages.push(
-        getMessageFromId({
-          messageId: 'wrongArgumentNumber',
-          values: {
-            fn: fn.name,
-            numArgs: max,
-            passedArgs: fn.args.length,
-          },
-          locations: fn.location,
-        })
-      );
-    } else if (fn.args.length > max) {
-      messages.push(
-        getMessageFromId({
-          messageId: 'wrongArgumentNumberTooMany',
-          values: {
-            fn: fn.name,
-            numArgs: max,
-            passedArgs: fn.args.length,
-            extraArgs: fn.args.length - max,
-          },
-          locations: fn.location,
-        })
-      );
-    } else {
-      messages.push(
-        getMessageFromId({
-          messageId: 'wrongArgumentNumberTooFew',
-          values: {
-            fn: fn.name,
-            numArgs: min,
-            passedArgs: fn.args.length,
-            missingArgs: min - fn.args.length,
-          },
-          locations: fn.location,
-        })
-      );
+    if (!this.licenseOk(this.definition.license)) {
+      return [errors.licenseRequired(this.fn, this.definition.license!)]; // TODO - maybe don't end validation here
```

Comment:
`this.definition.license!`
in reality this value can be undefined,  I'd like to avoid the on-null Assertion Operator in this case

I think it's better to be explicit
` if (this.definition.license && !this.licenseOk(this.definition.license)) {
      return [errors.licenseRequired(this.fn, this.definition.license)]; // TODO - maybe don't end validation here
    }`

`// TODO - maybe don't end validation here`

in the original logic I think  we just add the message to the message stack

`    const licenseMessages = validateFunctionLicense(fn, hasMinimumLicenseRequired);

Reply:
Thanks for confirming @bartoval !